### ISSUE TITLE
error behavior when no gpsd?

### ISSUE THREAD

**Issue by peribothron** [jmechnich/qtposition_gpsd#4]
[maybe this is related to patches in issue #2, but they do not compile for me]

if gpsd is *not* running, what property or signal should be checked?
PositionSource property *valid* seems to always be set to true. so it seems there is no way to detect non-active gpsd that way

> This property is true if the PositionSource object has acquired a valid backend plugin to provide data. If false, other methods on the PositionSource will have no effect.
> 
> Applications should check this property to determine whether positioning is  available and enabled on the runtime platform, and react accordingly.

to match mobile behaviour ("is location services running?"), maybe i would expect 
```
PositionSource.AccessError
```
or even
```
PositionSource.UnknownSourceError
```
but all i see is
```
PositionSource.NoError
```

here is a test code to try with or without gpsd running:
```
import QtQuick 2.8
import QtPositioning 5.5
//test plugin qtposition_gpsd-master/libqtposition_gpsd.so
//"valid" property

Rectangle {
  Text {
    anchors.centerIn: parent
    text: geoSource.valid? "source valid!" : "source invalid"
}

PositionSource {
  id: geoSource
  onSourceErrorChanged: {
    console.log("SourceErrorChanged: error code: " + sourceError)
  }
  Component.onCompleted: {
    console.log( "is PositionSource valid? " + valid )
    console.log( "is gpsd available? run: ps cax | grep gpsd" )
  }
}

}
```

**Comment by peribothron** [comment]
tldr:  the constructor 
```
QGeoPositionInfoSourceGpsd(QObject *parent)

```
will create an invalid position source if gpsd is not running. i do not know how to correct this.

the patches in #2 will compile with
```
CONFIG += debug

```
using this with a modified form of the qml code above shows the sequence of error signals:
```
$ qmlscene-qt5 badvalid.qml
GpsdMasterDevice::socketStateChanged state=host lookup
socket to gpsd opened
Created slave QBuffer(0x15ae9e0)
qml: PositionSource valid? true
qml:               active? false
qml:               req update interval [ms]? 2000
qml: SourceError? error code: 3
qml: PositionSource.NoError
qml: is gpsd available? run: ps cax | grep gpsd
qml: is gpsd port active? run: netstat -a | grep 2947
GpsdMasterDevice::socketStateChanged state=connecting
GpsdMasterDevice::socketStateChanged state=unconnected
GpsdMasterDevice::socketError error=connection refused
Destroyed slave QBuffer(0x15ae9e0)
Disconnecting from gpsd

```
modifying gpsdConnect to catch the inevitable failure can cause the socketError signal to appear before constructor returns:
```
  _socket->connectToHost(_hostname, _port);
  //if( !_socket->isOpen())
  int msecs = 2000;
  if (!_socket->waitForConnected(msecs)) 
  {
    qCritical() << "Could not open connection to gpsd in " << msecs;
    return false;
  }

```

if a c++ or qt guru can suggest how to modify the constructor to result in a position source that is not *valid* and has the appropriate *PositionSource.AccessError* i could try to submit a patch

**Comment by jmechnich** [comment]
Hi peribothron,

sorry for the late reply but I have finally managed to have a look.

I have submitted a few updates which might solve (at least part of) your issues. It was in fact the case that failing to connect to a GPSD instance was not handled correctly as it resulted in an open socket (I am not sure though if there was a subtle change between different Qt versions). However, as far as I can tell, there is no notification if the connection is lost or the GPSD instance is killed.
Let me know if the update helps or if there is still an open issue, I might be able to fix it.

Cheers,
Jörg

**Comment by jmechnich** [comment]
Part of your issue might be the fact that I used QNmeaPositionInfoSource as a base class which already implements some error handling which was good enough for me at that time. You might be able to implement your own error handling by emitting an QGeoPositionInfoSource::error(QGeoPositionInfoSource::Error positioningError) signal at the appropriate point and implementing a different QGeoPositionInfoSource::error() function. I did not have a closer look at that, though...

**Comment by jmechnich** [comment]
Closed due to inactivity.
